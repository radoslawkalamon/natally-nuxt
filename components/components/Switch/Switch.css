.switch {
  --switch-height: 24px;
  --switch-width: 48px;
  --switch-padding: 6px;
  --switch-checked-position: 24px;
  --switch-unchecked-position: 0;
  --switch-knob-size: 12px;
}

.switch__label-wrapper {
  align-items: center;
  display: flex;
  gap: var(--gap);
  margin-bottom: var(--gap);
}

.switch__checkbox {
  appearance: none;
  border-radius: var(--switch-height);
  box-shadow: inset 0 0 0 var(--border-size) var(--color-base);
  box-sizing: border-box;
  cursor: pointer;
  height: var(--switch-height);
  margin: 0;
  min-width: var(--switch-width);
  position: relative;
  transition: box-shadow var(--time-animation) ease, background-color var(--time-animation) ease;
  width: var(--switch-width);
}

.switch__checkbox::before {
  background-color: var(--color-base);
  border-radius: 50%;
  box-sizing: border-box;
  content: '';
  display: block;
  height: var(--switch-knob-size);
  left: var(--switch-padding);
  position: absolute;
  top: var(--switch-padding);
  transform: translateX(var(--switch-unchecked-position));
  transition: background-color var(--time-animation) ease, transform var(--time-animation) ease;
  width: var(--switch-knob-size);
  z-index: 1;
}

.switch__checkbox:checked {
  background-color: var(--color-accent);
  box-shadow: inset 0 0 0 var(--border-size) var(--color-border);
}

.switch__checkbox:checked::before {
  background-color: var(--color-off-white-1);
  transform: translateX(var(--switch-checked-position));
}

.switch__checkbox:disabled {
  box-shadow: inset 0 0 0 var(--border-size) var(--color-betwixt);
  cursor: not-allowed;
}

.switch__checkbox:disabled::before {
  background-color: var(--color-betwixt);
}

.switch__checkbox:disabled:checked {
  background-color: var(--color-betwixt);
  box-shadow: inset 0 0 0 var(--border-size) var(--color-border);
}

.switch__checkbox:disabled:checked::before {
  background-color: var(--color-off-white-1);
}

.switch__label {
  font-family: var(--font-family-sans);
  font-size: var(--font-size-header-3);
  font-weight: bold;
  line-height: var(--line-height-header);
  margin: 0;
}

.switch__label::after {
  color: var(--color-accent);
  content: var(--content-after-header);
}

.switch__description {
  width: 100%;
}
