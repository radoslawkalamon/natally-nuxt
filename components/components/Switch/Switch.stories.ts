import type { Meta, Story } from '@storybook/vue'
import ComponentsSwitch from '@/components/components/Switch/Switch.vue'

const meta: Meta = {
  title: 'Components / Switch',
  component: ComponentsSwitch
}
export default meta

const Template: Story = (_args, { argTypes }) => ({
  components: {
    ComponentsSwitch
  },
  props: Object.keys(argTypes),
  template: `<ComponentsSwitch v-bind="$props">
    <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer ultrices pulvinar ligula, sit amet dignissim turpis molestie malesuada. Mauris cursus velit augue, at lobortis ex sagittis nec. Cras ac pulvinar est, nec varius risus. Donec feugiat, tellus ac mattis sodales, nisi tortor congue lorem, nec varius magna lorem quis nibh. Donec lacinia, ligula quis rutrum ornare, urna nulla varius enim, eu venenatis orci lacus id tellus. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Sed eu quam nibh. Morbi in volutpat eros. Vivamus pulvinar non nulla ut imperdiet. Etiam accumsan aliquet vehicula.</p>
  </ComponentsSwitch>`
})

export const Checked: Story = Template.bind({})
Checked.args = {
  checked: true,
  disabled: false,
  label: 'Switch label',
  name: 'switch-name'
}

export const Unchecked: Story = Template.bind({})
Unchecked.args = {
  checked: false,
  disabled: false,
  label: 'Switch label',
  name: 'switch-name'
}

export const Disabled: Story = Template.bind({})
Disabled.args = {
  checked: true,
  disabled: true,
  label: 'Switch label',
  name: 'switch-name'
}
