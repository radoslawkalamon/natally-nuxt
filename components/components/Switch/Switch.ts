import Vue from 'vue'

export default Vue.extend({
  name: 'ComponentsSwitch',
  components: {
    WrappersText: () => import(/* webpackChunkName: "wrappers-text" */'@/components/wrappers/Text/Text.vue')
  },
  inheritAttrs: false,
  props: {
    label: {
      type: String,
      required: true
    },
    name: {
      type: String,
      required: true
    }
  }
})
