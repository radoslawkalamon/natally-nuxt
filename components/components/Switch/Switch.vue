<template>
  <div
    class="switch"
    data-test="components-switch"
  >
    <label
      class="switch__label-wrapper"
      :for="name"
    >
      <input
        :id="name"
        class="switch__checkbox"
        data-test="components-switch-input"
        type="checkbox"
        :name="name"
        v-bind="$attrs"
        v-on="$listeners"
      >
      <span
        class="switch__label"
        v-text="label"
      />
    </label>
    <WrappersText
      v-if="$slots.default"
      class="switch__description"
      data-test="components-switch-description"
    >
      <slot />
    </WrappersText>
  </div>
</template>

<script src="./Switch.ts" lang="ts"></script>
<style src="./Switch.css" scoped></style>
