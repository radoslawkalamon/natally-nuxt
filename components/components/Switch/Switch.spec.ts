import { mount } from '@vue/test-utils'
import Switch from './Switch.vue'
import { shallHaveStringProp, shallHaveSlot, shallHideSlotWhenNoChildren, shallRender } from '@/devtools/jest.shared.spec'

describe('Components / Switch', () => {
  const defaultOptions = {
    propsData: {
      checked: false,
      label: 'Test label',
      name: 'test-name'
    },
    stubs: {
      WrappersText: {
        template: '<div><slot /></div>'
      }
    }
  }

  shallRender(Switch, defaultOptions)
  shallHaveStringProp(Switch, 'label', defaultOptions)
  shallHideSlotWhenNoChildren(Switch, 'default', 'components-switch-description', defaultOptions)
  shallHaveSlot(Switch, 'default', defaultOptions)

  test('shall have for="test-name" and name="test-name" elements', () => {
    const wrapper = mount(Switch, defaultOptions)
    const forElement = wrapper.find('[for="test-name"]')
    const nameElement = wrapper.find('[name="test-name"]')
    expect(forElement.exists()).toBeTruthy()
    expect(nameElement.exists()).toBeTruthy()
  })
})
