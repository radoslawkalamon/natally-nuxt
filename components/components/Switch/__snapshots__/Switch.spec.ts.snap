// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components / Switch Integration shall pass integration sanity test | no slot 1`] = `
"<div class="switch"><label for="test-name" class="switch__label-wrapper"><input id="test-name" type="checkbox" name="test-name" class="switch__checkbox">
    <div class="switch__fake-checkbox"></div> <span class="switch__label">Test label</span>
  </label>
  <!---->
</div>"
`;

exports[`Components / Switch Integration shall pass integration sanity test | with slot 1`] = `
"<div class="switch"><label for="test-name" class="switch__label-wrapper"><input id="test-name" type="checkbox" name="test-name" class="switch__checkbox">
    <div class="switch__fake-checkbox"></div> <span class="switch__label">Test label</span>
  </label>
  <div class="switch__description text">Default slot test</div>
</div>"
`;

exports[`Components / Switch Unit shall pass unit sanity test | no slot 1`] = `
"<div class="switch"><label for="test-name" class="switch__label-wrapper"><input id="test-name" type="checkbox" name="test-name" class="switch__checkbox">
    <div class="switch__fake-checkbox"></div> <span class="switch__label">Test label</span>
  </label>
  <!---->
</div>"
`;

exports[`Components / Switch Unit shall pass unit sanity test | with slot 1`] = `
"<div class="switch"><label for="test-name" class="switch__label-wrapper"><input id="test-name" type="checkbox" name="test-name" class="switch__checkbox">
    <div class="switch__fake-checkbox"></div> <span class="switch__label">Test label</span>
  </label>
  <wrapperstext-stub class="switch__description">Default slot test</wrapperstext-stub>
</div>"
`;
